Implements compatibility api

rbans96 opened the pull request:
## Description
Implements the _schema/compatibility to return min and max compatible schema versions from the server
Adds stored procedure SelectMaxSupportedSchemaVersion

## Related issues
User Story [AB#72833](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/72833).

## Testing
Describe how this change was tested.
Unit test and manual tests
![image](https://user-images.githubusercontent.com/57157506/76358823-59110600-62d7-11ea-906b-3349fe07ded1.png)

brendankowitz commented on src/Microsoft.Health.Fhir.SqlServer.Api/Controllers/SchemaController.cs:
```diff
@@ -82,7 +87,15 @@ public ActionResult Compatibility()
 
-            throw new NotImplementedException(Resources.CompatibilityNotImplemented);
+            var minVersion = (int)_schemaInformation.MinimumSupportedVersion;
+            var compatibilityVersions = new Dictionary<string, object> { { "min", minVersion } };
```
I think we should create POCO models going forward

brendankowitz commented on src/Microsoft.Health.Fhir.Core/Features/Schema/IQueryProcessor.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Microsoft.Health.Fhir.Core.Features.Schema
+{
+    public interface IQueryProcessor
```
Using a Mediatr handler is the common pattern we currently have for calling functionality from the controller, (maybe `LatestSchemaVersionHandler`).

brendankowitz commented on src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/2.diff.sql:
```diff
@@ -237,3 +237,28 @@ AS
+--  RETURNS
+--      The highest version
+--
+CREATE PROCEDURE dbo.SelectMaxSupportedSchemaVersion
```
Should be in V3?

rbans96 replied:
Yes, since 3.sql was not merged, so kept it in V2. I will move it to V3.

rbans96 commented on src/Microsoft.Health.Fhir.Core/Features/Schema/LatestSchemaVersionHandler.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            EnsureArg.IsNotNull(request, nameof(request));
+
+            if (await _fhirAuthorizationService.CheckAccess(DataActions.SchemaMigration) != DataActions.SchemaMigration)
```
Due to some reason the api is throwing Authorization exception. I have added new enum SchemaMigration in DataAccess. Not sure what am I missing here :( Please suggest.

brendankowitz replied:
@johnstairs any guidance here?
I think the controller action has `AllowAnonymous` so it will probably not be able to check roles/access for unauth'd users

rbans96 replied:
I discussed with John this morning, since schema migration will be run by admin by hitting few commands on tool, we might not need authorization check for schema migration specific apis. As of now i have removed this check. But I am all ears for any suggestions.

brandonpollett replied:
I agree that the AuthN/AuthZ isn't needed for these APIs.

brendankowitz commented on src/Microsoft.Health.Fhir.Core/Features/Operations/IFhirOperationDataStore.cs:
```diff
@@ -56,5 +56,13 @@ public interface IFhirOperationDataStore
+        /// <param name="maxVersion">The maximum schema version specified by code</param>
+        /// /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The latest supported schema version from server.</returns>
+        Task<int> GetLatestCompatibleVersionAsync(int maxVersion, CancellationToken cancellationToken);
```
In general things are looking good, some thoughts:
* Perhaps break this into its own interface `ISchemaMigrationDataStore`
* Think about how you can gracefully have the CosmosDb implementation return an unsupported status

rbans96 replied:
yes it was in my mind to separate this out. But I know Brandon is refactoring lot of files to break components to get sql out. So I am looking forward for his PR too. Anyways I will update it. Thank you!

brandonpollett replied:
I think we should limit the schema stuff to the SQL packages. I'm not sure it has broad applicability to the Cosmos implementation and also the `_schema` routes only exist in the SQL packages.

rbans96 replied:
Addressed it.

brandonpollett commented on src/Microsoft.Health.Fhir.Core/Features/Schema/ISchemaMigrationDataStore.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Microsoft.Health.Fhir.Core.Features.Schema
+{
+    public interface ISchemaMigrationDataStore
```
Why does this live here instead of in the SQL packages?

brandonpollett commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Sql/SchemaTests.cs:
```diff
@@ -54,11 +54,16 @@ public async Task WhenRequestingAvailable_GivenAServerThatHasSchemas_JsonShouldB
+        [Fact]
+        public async Task WhenRequestingSchema_GivenGetMethodAndCompatibilityPath_TheServerShouldReturnSuccess()
+        {
+            await SendAndVerifyStatusCode(HttpMethod.Get, "_schema/compatibility", HttpStatusCode.OK);
```
Probably want to verify that it returns the correct thing, too.

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer.Api/Controllers/SchemaController.cs:
```diff
@@ -78,11 +83,15 @@ public FileContentResult SqlScript(int id)
-            throw new NotImplementedException(Resources.CompatibilityNotImplemented);
+            int maxVersion = (int)_schemaInformation.MaximumSupportedVersion;
+            int minVersion = (int)_schemaInformation.MinimumSupportedVersion;
+            var compatibleResponse = await _mediator.GetCompatibleVersionAsync(minVersion, maxVersion, HttpContext.RequestAborted);
```
This will vary based on the instance that you hit. I would think that it would be more helpful to use the current version in the DB along with the information we're recording in the instance table.

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/3.sql:
```diff
@@ -82,6 +82,31 @@ BEGIN
+
+    SELECT MAX(Version)
+    FROM SchemaVersion
+    WHERE Status = 'complete' AND Version <= @version
```
This just seems like it's basically giving the current version in most cases unless you happen to hit an instance that hasn't updated to the latest version.

rbans96 replied:
Updated the query, please review and let me know your thoughts5

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer.Api.UnitTests/Controllers/SchemaControllerTests.cs:
```diff
@@ -58,11 +61,5 @@ public void GivenACurrentVersiontRequest_WhenNotImplemented_ThenNotImplementedSh
         }
-
-        [Fact]
-        public void GivenACompatibilityRequest_WhenNotImplemented_ThenNotImplementedShouldBeThrown()
```
Since we're removing the error test, do we want to add a success test?

rbans96 replied:
Test is added in the  LatestVersionSchemaHandler for success scenario  and there is no particular check in controller before it gets processed by the handler. Let me know if we still need one in controller

brandonpollett commented on src/Microsoft.Health.Fhir.Shared.Api.UnitTests/Features/Security/RequestHandlerCheckAccessTests.cs:
```diff
@@ -50,7 +50,7 @@ void Dfs(Assembly a)
 
             return assemblies.SelectMany(a => a.GetTypes())
-                .Where(t => t.IsClass && !t.IsAbstract && t.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IRequestHandler<,>)))
+                .Where(t => t.IsClass && string.Compare("LatestSchemaVersionHandler", t.Name, StringComparison.OrdinalIgnoreCase) != 0 && !t.IsAbstract && t.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IRequestHandler<,>)))
```
Shouldn't be needed right?

brendankowitz replied:
There is also an `_exceptions` list above, would be a better place for this

rbans96 replied:
Its not needed as the handler is now moved to sql package

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer.Api/Controllers/SchemaController.cs:
```diff
@@ -24,8 +27,9 @@ public class SchemaController : Controller
+        private readonly IMediator _mediator;
 
-        public SchemaController(SchemaInformation schemaInformation, IUrlResolver urlResolver, ILogger<SchemaController> logger)
+        public SchemaController(SchemaInformation schemaInformation, IUrlResolver urlResolver, ILogger<SchemaController> logger, IMediator mediator)
```
nit: convention is that logger should be last item

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer.Api/Controllers/SchemaController.cs:
```diff
@@ -78,11 +83,14 @@ public FileContentResult SqlScript(int id)
             _logger.LogInformation("Attempting to get compatibility");
 
-            throw new NotImplementedException(Resources.CompatibilityNotImplemented);
+            int minVersion = (int)_schemaInformation.MinimumSupportedVersion;
```
This is going to vary based on the instance that you happen to hit if the code packages are different.

rbans96 replied:
Updated in the last commit to fetch minVersion too from InstanceSchema table

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer/Features/Exceptions/RecordNotFoundException.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Exceptions
+{
+    public class RecordNotFoundException : FhirException
```
This probably shouldn't extend `FhirException` since it doesn't actually deal with FHIR.

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Extensions/SchemaMediatorExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+
+            var request = new GetCompatibilityVersionRequest(minVersion);
+
+            GetCompatibilityVersionResponse response = await mediator.Send(request, cancellationToken);
```
Could probably just `return await mediator.Send(request, cancellationToken);`

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer/Features/Schema/ISchemaMigrationDataStore.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Schema
+{
+    public interface ISchemaMigrationDataStore
```
Maybe just `ISchemaDataStore` or `ISchemaVersionDataStore`. Thoughts?

brandonpollett commented on src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerSchemaMigrationDataStore.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public async Task<int> GetLatestCompatibleVersionAsync(CancellationToken cancellationToken)
+        {
+            using (SqlConnectionWrapper sqlConnectionWrapper = _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapper(true))
```
Does this actually need to enlist in a transaction? Is there an existing transaction?

brendankowitz commented on src/Microsoft.Health.Fhir.SqlServer.Api/Controllers/SchemaController.cs:
```diff
@@ -34,6 +38,7 @@ public SchemaController(SchemaInformation schemaInformation, IUrlResolver urlRes
             _schemaInformation = schemaInformation;
             _urlResolver = urlResolver;
             _logger = logger;
+            _mediator = mediator;
```
EnsureArg for mediator

rbans96 commented:
This PR changes and addressed feedback are present in current version api PR because of the similar changes required.
Please refer https://github.com/microsoft/fhir-server/pull/938 for the changes
